kale/embed/gripnet.py

```diff
@@ -241,3 +245,294 @@ def forward(
+        self.in_channels = in_channels
+        self.out_channels = setting.inter_agg_dim[-1]
+
+        self.num_edge_type = num_edge_type
+        self.if_multirelational = 1 if num_edge_type > 1 else 0
+        self.if_start_svertex = if_start_svertex
+        self.setting = setting
+
+        self.__init_inter_feat_layer__()
+        self.__init_inter_agg_layer__()
+
+    def __init_inter_feat_layer__(self):
+        """internal feature layer"""
+
+        self.embedding = torch.nn.Parameter(torch.Tensor(self.in_channels, self.setting.inter_feat_dim))
+        self.embedding.requires_grad = True
+
+        # reset parameters to be normally distributed
+        self.embedding.data.normal_()
+
+    def __init_inter_agg_layer__(self):
+        """internal aggregation layers"""
+
+        # compute the dim of input of the first internal aggregation layer
+        self.in_agg_dim = self.setting.inter_feat_dim
+        if not self.if_start_svertex:
+            assert self.setting.mode in [
+                "cat",
+                "add",
+            ], f"The mode {self.setting.mode} is not supported. Please use cat or add."
+
+            if self.setting.mode == "cat":
+                assert self.setting.exter_agg_dim, "The exter_agg_dim is not set."
+                self.in_agg_dim += sum(self.setting.exter_agg_dim.values())
+            else:
+                tmp = set([self.in_agg_dim] + list(self.setting.exter_agg_dim.values()))
+                assert len(tmp) == 1, "The in_agg_dim should be the same as any element in exter_agg_dim."
+
+        # create and initialize the internal aggregation layers
+        self.num_inter_agg_layer = len(self.setting.inter_agg_dim)
+        tmp_dim = [self.in_agg_dim] + self.setting.inter_agg_dim
+
+        if self.setting.if_catout:
+            self.out_channels = sum(tmp_dim)
+
+        if self.if_multirelational:
+            # using RGCN if there are multiple edge types
+            after_relu = [False if i == 0 else True for i in range(self.num_inter_agg_layer)]
+            self.inter_agg_layers = torch.nn.ModuleList(
+                [
+                    RGCNEncoderLayer(
+                        tmp_dim[i], tmp_dim[i + 1], self.num_edge_type, self.setting.num_bases, after_relu[i]
+                    )
+                    for i in range(self.num_inter_agg_layer)
+                ]
+            )
+        else:
+            # using GCN if there is only one edge type
+            self.inter_agg_layers = torch.nn.ModuleList(
+                [GCNEncoderLayer(tmp_dim[i], tmp_dim[i + 1], cached=True) for i in range(self.num_inter_agg_layer)]
+            )
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        edge_index: torch.Tensor,
+        edge_type: torch.Tensor = None,
+        range_list: torch.Tensor = None,
+        edge_weight: torch.Tensor = None,
+    ) -> torch.Tensor:
+        r"""
+        Args:
+            x (torch.Tensor): the input node feature embedding.
```

Comment:
can you give x a more descriptive name?

Reply:
It would be better to use `x` here, as it is consistent with the `forward()` function of the `*.nn.modules` in PyTorch and PyG.